website/server/libs/spells.js

```diff
@@ -163,14 +164,30 @@ async function castSpell (req, res, { isV3 = false }) {
       task: results[1],
     });
   } else if (targetType === 'self') {
-    await castSelfSpell(req, user, spell, quantity);
-
-    let userToJson = user;
-    if (isV3) userToJson = await userToJson.toJSONWithInbox();
-
-    res.respond(200, {
-      user: userToJson,
-    });
+    // Check if chilling frost or stealh skill has been previously casted or not.
+    // See #12361 for more details.
+    const spellName = spell.key;
+    // const incompleteDailiesDue = getUnfilteredTaskList('daily').filter(daily => (daily.completed == false && daily.isDue === true)).length;
```

Comment:
@paglias Facing an issue here. When I try to use the `incompleteDailiesDue` function, it imports it from `website/client/src/store/getters/tasks.js` as it has been exported from there. 

However, after doing so, I keep getting this error.
```
/home/jalansh/Desktop/habitica/website/client/src/store/getters/tasks.js:1
import sortBy from 'lodash/sortBy';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
As of now, I have commented out the logic for the Stealth skill. I referred this method's usage in the PR #12255 . Do you know what might cause this?

As another approach, we can implement the same method or get the incomplete and due dailies for the user in spells.js to avoid this error.

Reply:
incompleteDailiesDue is current part of the client side code, to use it on the server it'll have to be duplicated in a new function here or be moved to /common/script so that it can be used in both places

Reply:
but in this case I think the code would be different so I'd make a new function

Reply:
Thanks for the feedback, @paglias ! That helped. I added code for getting incomplete and due dailies and it works.

I also need the logic in `website/common/script/content/spells.js` to avoid the deduction in `MP`. I wrote the same implementation there as well. But I am facing a console error in my browser saying (and the website does not load) - 
```
Uncaught Error: Cannot find module './script/index'
    webpackEmptyContext common sync:2
    <anonymous> index.js:10
    js app.js:1017
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> task.js:41
    js app.js:2493
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> spells.js:19
    js app.js:1629
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> index.js:24
    js app.js:1557
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> index.js:9
    js app.js:16252
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> analytics.js:23
    js app.js:15580
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> cjs.js:17
    ./node_modules/cache-loader/dist/cjs.js?!./node_modules/babel-loader/lib/index.js!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./src/components/header/menu.vue?vue&type=script&lang=js& app.js:4539
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> menu.vue:1
    ./src/components/header/menu.vue?vue&type=script&lang=js& app.js:12460
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> menu.vue:1
    vue app.js:12448
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> cjs.js:37
    ./node_modules/cache-loader/dist/cjs.js?!./node_modules/babel-loader/lib/index.js!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./src/app.vue?vue&type=script&lang=js& app.js:4012
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> app.vue:1
    ./src/app.vue?vue&type=script&lang=js& app.js:8156
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> app.vue:1
    vue app.js:8144
    __webpack_require__ app.js:854
    fn app.js:151
    <anonymous> main.js:12
    js app.js:15784
    __webpack_require__ app.js:854
    fn app.js:151
    1 app.js:16265
    __webpack_require__ app.js:854
    checkDeferredModules app.js:46
    <anonymous> app.js:994
    <anonymous> app.js:997
app.js line 1006 > eval:2:10
```
I tried fixing it but was not able to. Can we not use a method of the server side code in `/common/script`? Is there any other method by which I can write this?

Reply:
`common` files can import from one another, and server and client code can use common code, but common can't use server- or client-only code. Can you walk me through what you're trying to solve and why the server code is needed? I'm not spotting it on a look through the PR as it stands.

Reply:
@SabreCat Okay, I understand now. I need to get the number of dailies which are incomplete and due. For that, I added the following code - 
```
const incompleteDailiesDue = Tasks.Task.find({
        userId: user._id,
        type: 'daily',
        $and: [
          { type: 'daily', completed: false, isDue: true },
        ],
      }).exec();
```
This works perfectly on the server side. I also need the list in `common` to avoid deducting `MP`. But I won't be able to import the `Tasks` model from server according to what you said. 

Hence, it is best if we implement the function in `common` so that we can use it on the server side as well. but how can I use the `Tasks` model here? Is there any other way for getting the list of incomplete and due dailies? Does this clarify the problem?

Reply:
Aha, thanks for the clarification, I understand the trouble now!

Probably the most appropriate place to carry out a calculation like that would be in `castSpell` in `website/server/libs/spells.js`... but that'd require making a special case for `stealth` there, or having e.g. an `abortFunction` property in the spell content that provides code for the server to run, neither of which feels great. @paglias what do you think?

Reply:
I think the only way to do this is to have a special case for `stealth` as @SabreCat suggested in `website/server/libs/spells.js ` that will return an error in case the skill should not be casted, the code on the client can remain the same. If it turns out that in future a similar exception is needed for more skills we can look into make it more generic but for now I don't think it's necessary

Reply:
Thanks for the reply, @SabreCat and @paglias !

I am already throwing a `BadRequest` for the repeated `Stealth` skill. 
```
else if (spellName === 'stealth' && user.stats.buffs.stealth >= incompleteDailiesDue.length) {
      throw new BadRequest(res.t('spellRogueStealthMaxedOut'));
    } 
```
By keeping the client-side code same, are you saying that we are letting the `MP` deduct? Do I need to do include additional code in the `castSpell` method in `website/server/libs/spells.js` for avoiding the change in `MP`? Can you please clarify?

Reply:
> By keeping the client-side code same, are you saying that we are letting the MP deduct?

No I meant that the client side already has a check to prevent using `stealth`  when `user.stats.buffs.stealth >= incompleteDailiesDue.length` so it doesn't require any change

Reply:
Ah, got it. I noticed that `MP` was deducted on a repeating cast, but that was happening only in the UI. The `MP` was not deducted when I reloaded the page. Is this expected? Should this be changed?

Reply:
Hmm, that does sound incorrect... if the code @paglias mentions were working correctly, then you'd see the alert about not needing to use Stealth again, not a deduction of MP that is later reverted.

Reply:
@SabreCat The alert will be visible when my changes are merged with the master branch. The API will send a `BadRequest` which will lead to an error. The tooltip (see #12255 ) is also visible when we try to recast Stealth or Chilling Frost. 

What I am saying is, the UI shows that `MP` is deducted and that is why I initially included the logic to avoid `MP` deduction in `website/common/script/content/spells.js`. However, `MP` is not actually deducted in the database according to what I observed on my local install. Hence, it works fine as @paglias mentioned, but I guess ideally the UI should not show the deduction. What do you think? Please check this from your end as well.

Reply:
OK, I see what's going on with the client--we have logic to show the "you don't need to cast this again" tooltip, but it doesn't actually prevent the cast. For that we'd need to e.g. change line 45 of website/client/src/components/tasks/spells.vue to `@click="!spellDisabled(key) ? castStart(skill) : null"`

So if you'd like to fix that, you can, but it's fine to keep this PR's scope server-side! In which case, the spurious -MP result that's reverted after sync is probably acceptable.

I am seeing that the error message you get for attempting to overuse Stealth talks about streaks being frozen, which is the Chilling Frost description, though.

Reply:
If we could fix the MP deduction in the client side in this PR it would be great, I thought it was already fixed in the previous PR but it doesn't look like so

Reply:
Yes, sure. I will make the changes that @SabreCat suggested in my next commit. While we are it, should I also add the wording change that @veeeeeee suggested in #12361 for both `Stealth` and `Chilling Frost` skills? And if we are keeping the same message for both the skills, then should we keep a common JSON key for both the skills instead of two different keys?

Please note that I have added the test case for both the cases. Please have a look at it and let me know if something needs to be changed.

Reply:
Ah yes, good catch on V's suggested wording. One string token used in both cases should be fine!